core/lib/zksync_core/src/api_server/web3/tests/ws.rs

```diff
@@ -0,0 +1,694 @@
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn log_filter_changes() {
+    test_ws_server(LogFilterChanges).await;
+}
+
+#[derive(Debug)]
+struct LogFilterChangesWithBlockBoundaries;
+
+#[async_trait]
+impl WsTest for LogFilterChangesWithBlockBoundaries {
+    async fn test(&self, client: &WsClient, pool: &ConnectionPool) -> anyhow::Result<()> {
+        let lower_bound_filter = Filter {
+            from_block: Some(api::BlockNumber::Number(2.into())),
+            ..Filter::default()
+        };
+        let lower_bound_filter_id = client.new_filter(lower_bound_filter).await?;
+        let upper_bound_filter = Filter {
+            to_block: Some(api::BlockNumber::Number(1.into())),
+            ..Filter::default()
+        };
+        let upper_bound_filter_id = client.new_filter(upper_bound_filter).await?;
+        let bounded_filter = Filter {
+            from_block: Some(api::BlockNumber::Number(1.into())),
+            to_block: Some(api::BlockNumber::Number(1.into())),
+            ..Filter::default()
+        };
+        let bounded_filter_id = client.new_filter(bounded_filter).await?;
+
+        let mut storage = pool.access_storage().await?;
+        let (_, events) = LogSubscriptions::update_storage(&mut storage, 1, 0).await?;
+        drop(storage);
+        let events: Vec<_> = events.iter().collect();
+
+        let lower_bound_logs = client.get_filter_changes(lower_bound_filter_id).await?;
+        assert_matches!(
+            lower_bound_logs,
+            FilterChanges::Hashes(hashes) if hashes.is_empty()
+        );
+        // ^ Since `FilterChanges` is serialized w/o a tag, an empty array will be deserialized
+        // as `Hashes(_)` (the first declared variant).
+
+        let upper_bound_logs = client.get_filter_changes(upper_bound_filter_id).await?;
+        let FilterChanges::Logs(upper_bound_logs) = upper_bound_logs else {
+            panic!("Unexpected getFilterChanges output: {:?}", upper_bound_logs);
+        };
+        assert_logs_match(&upper_bound_logs, &events);
+        let bounded_logs = client.get_filter_changes(bounded_filter_id).await?;
+        let FilterChanges::Logs(bounded_logs) = bounded_logs else {
+            panic!("Unexpected getFilterChanges output: {:?}", bounded_logs);
+        };
+        assert_eq!(bounded_logs, upper_bound_logs);
+
+        // Add another miniblock with events to the storage.
+        let mut storage = pool.access_storage().await?;
+        let (_, new_events) = LogSubscriptions::update_storage(&mut storage, 2, 4).await?;
+        drop(storage);
+        let new_events: Vec<_> = new_events.iter().collect();
+
+        let lower_bound_logs = client.get_filter_changes(lower_bound_filter_id).await?;
+        let FilterChanges::Logs(lower_bound_logs) = lower_bound_logs else {
+            panic!("Unexpected getFilterChanges output: {:?}", lower_bound_logs);
+        };
+        assert_logs_match(&lower_bound_logs, &new_events);
+
+        // FIXME: is this expected?
```

Comment:
A fixme should either be figured out before merging, or probably a new tracking issue has to be created and referenced here.

Reply:
Yep, it should be discussed whether this `getFilterChanges` behavior is expected (IMO, it's not, but I don't have experience with `getFilterChanges` in Ethereum), and creating a task to fix it if it's not.

Reply:
> Polling method for a filter, which returns an array of logs which occurred since last poll.

Yeah we should create a task

Reply:
I think that the bug was introduced in #526 

https://github.com/matter-labs/zksync-era/blob/925da43c3b85b4818cb910f286b0f69ae118d692/core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs#L862-L868

Reply:
This behavior most certainly wasn't introduced in #526. The initial version of this PR with the tests predates #526 being merged; #526 just reproduces the behavior that already existed (just slightly refactors it). This is what the FIXME is about – investigating whether the existing behavior is correct, and fixing it if it's not.